elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/PredicateAnalyzer.java

```diff
@@ -0,0 +1,966 @@
+import static org.apache.calcite.adapter.elasticsearch.QueryBuilders.existsQuery;
+import static org.apache.calcite.adapter.elasticsearch.QueryBuilders.rangeQuery;
+import static org.apache.calcite.adapter.elasticsearch.QueryBuilders.regexpQuery;
+import static org.apache.calcite.adapter.elasticsearch.QueryBuilders.termQuery;
+
+import static java.lang.String.format;
+
+/**
+ * Query predicate analyzer. Uses visitor pattern to traverse existing expression
+ * and convert it to {@link QueryBuilder}.
+ *
+ * <p>Major part of this class have been copied from
+ * <a href="https://www.dremio.com/">dremio</a> ES adapter
+ * (thanks to their team for improving calcite-ES integration).
+ */
+class PredicateAnalyzer {
+
+  /**
+   * Internal exception
+   */
+  @SuppressWarnings("serial")
+  private static final class PredicateAnalyzerException extends RuntimeException {
+
+    PredicateAnalyzerException(String message) {
+      super(message);
+    }
+
+    PredicateAnalyzerException(Throwable cause) {
+      super(cause);
+    }
+  }
+
+  private PredicateAnalyzer() {}
+
+  /**
+   * Walks the expression tree, attempting to convert the entire tree into
+   * an equivalent Elasticsearch query filter. If an error occurs, or if it
+   * is determined that the expression cannot be converted, an exception is
+   * thrown and an error message logged.
+   *
+   * <p>Callers should catch ExpressionNotAnalyzableException
+   * and fall back to not using push-down filters.
+   *
+   * @param expression expression to analyze
+   * @return search query which can be used to query ES cluster
+   * @throws ExpressionNotAnalyzableException when expression can't processed by this analyzer
+   */
+  static QueryBuilder analyze(RexNode expression) throws ExpressionNotAnalyzableException {
+    Objects.requireNonNull(expression, "expression");
+    try {
+      // does ES
+      QueryExpression e = (QueryExpression) expression.accept(new Visitor());
+
+      if (e != null && e.isPartial()) {
+        throw new UnsupportedOperationException("Can't handle partial QueryExpression: " + e);
+      }
+      return e != null ? e.builder() : null;
+    } catch (Throwable e) {
+      Throwables.propagateIfPossible(e, UnsupportedOperationException.class);
+      throw new ExpressionNotAnalyzableException("Can't convert " + expression, e);
+    }
+  }
+
+  /**
+   * Converts expressions of the form NOT(LIKE(...)) into NOT_LIKE(...)
+   */
+  private static class NotLikeConverter extends RexShuttle {
+    final RexBuilder rexBuilder;
+
+    NotLikeConverter(RexBuilder rexBuilder) {
+      this.rexBuilder = rexBuilder;
+    }
+
+    @Override public RexNode visitCall(RexCall call) {
+      if (call.getOperator().getKind() == SqlKind.NOT) {
+        RexNode child = call.getOperands().get(0);
+        if (child.getKind() == SqlKind.LIKE) {
+          List<RexNode> operands = FluentIterable.from(((RexCall) child).getOperands())
+              .transform(new Function<RexNode, RexNode>() {
```

Comment:
For new code, it would be good to stick to Java 8 lambda syntax now that the patch from @julianhyde has landed. That is `rexNode -> rexNode.accept(NotLikeConverter.this)`.